pkg/api/componentreadiness/test_details.go

```diff
@@ -25,55 +25,195 @@ import (
 
 func GetTestDetails(ctx context.Context, client *bigquery.Client, dbc *db.DB, reqOptions crtype.RequestOptions,
 ) (crtype.ReportTestDetails, []error) {
-	generator := ComponentReportGenerator{
-		client:     client,
-		dbc:        dbc,
-		ReqOptions: reqOptions,
-	}
+	generator := NewComponentReportGenerator(client, reqOptions, dbc, nil)
 	if os.Getenv("DEV_MODE") == "1" {
 		return generator.GenerateTestDetailsReport(ctx)
 	}
 
-	return api.GetDataFromCacheOrGenerate[crtype.ReportTestDetails](
+	report, errs := api.GetDataFromCacheOrGenerate[crtype.ReportTestDetails](
 		ctx,
 		generator.client.Cache,
 		generator.ReqOptions.CacheOption,
-		generator.GetComponentReportCacheKey(ctx, "TestDetailsReport~"),
+		api.GetPrefixedCacheKey("TestDetailsReport~", generator.GetCacheKey(ctx)),
 		generator.GenerateTestDetailsReport,
 		crtype.ReportTestDetails{})
+	if len(errs) > 0 {
+		return report, errs
+	}
+
+	err := generator.PostAnalysisTestDetails(&report)
+	if err != nil {
+		return report, []error{err}
+	}
+
+	return report, []error{}
 }
 
-func (c *ComponentReportGenerator) GenerateTestDetailsReport(ctx context.Context) (crtype.ReportTestDetails, []error) {
-	c.initializeMiddleware()
+// PostAnalysisTestDetails runs the PostAnalysis method for all middleware on this test details report.
+// This is done outside the caching mechanism so we can load fresh data from our db (which is fast and cheap),
+// and inject it into an expensive / slow report without recalculating everything.
+func (c *ComponentReportGenerator) PostAnalysisTestDetails(report *crtype.ReportTestDetails) error {
 
-	if c.ReqOptions.TestIDOption.TestID == "" {
-		return crtype.ReportTestDetails{}, []error{fmt.Errorf("test_id has to be defined for test details")}
-	}
-	for _, v := range c.ReqOptions.VariantOption.DBGroupBy.List() {
-		if _, ok := c.ReqOptions.VariantOption.RequestedVariants[v]; !ok {
-			return crtype.ReportTestDetails{}, []error{fmt.Errorf("all dbGroupBy variants have to be defined for test details: %s is missing in %v", v, c.ReqOptions.VariantOption.RequestedVariants)}
+	// Give middleware their chance to adjust the result
+	for _, mw := range c.middlewares {
+		testKey := crtype.ReportTestIdentification{
+			RowIdentification:    report.RowIdentification,
+			ColumnIdentification: report.ColumnIdentification,
+		}
+		for ai := range report.Analyses {
```

Comment:
why is this var named `ai`? analysis index?

Reply:
Yeah, maybe too overloaded :) I'll change it